File: concepts/08_async_and_concurrency/02_scheduling_tasks_microtasks_and_futures/lua/test_02_ready_queue_boundaries.lua
-- Common question: when does newly ready work run relative to current work?
-- Inputs: an explicit FIFO queue, tasks that enqueue tasks, and coroutine resumptions.
-- Observations: queue policy belongs to the host, current callback completion, and explicit ready boundaries.
-- polyglot-family: async_and_concurrency
-- polyglot-concept: scheduling_tasks_microtasks_and_futures
-- polyglot-related: languages/lua/language/08_coroutines_and_gc/test_057_coroutine_resume_and_results.lua

local t = require("support.assertions")
local queue = {}
local events = {}
local function schedule(callback)
    queue[#queue + 1] = callback
end

schedule(function()
    events[#events + 1] = "outer:start"
    schedule(function() events[#events + 1] = "inner" end)
    events[#events + 1] = "outer:end"
end)
schedule(function() events[#events + 1] = "peer" end)

while #queue > 0 do
    local callback = table.remove(queue, 1)
    callback()
end
t.equal(table.concat(events, ","), "outer:start,outer:end,peer,inner")

-- 该队列是宿主策略示例，不是 Lua 标准 event loop、future 或 microtask API。
t.done()

File: concepts/08_async_and_concurrency/02_scheduling_tasks_microtasks_and_futures/ruby/test_02_ready_queue_boundaries.rb
# 共同问题：当前 callback 新增的 ready work 相对当前工作何时执行。
# 输入：显式 FIFO Queue、会继续 enqueue 的任务和循环；观察：当前 callback 完成后才取下一个任务。
# polyglot-family: async_and_concurrency
# polyglot-concept: scheduling_tasks_microtasks_and_futures
# polyglot-related: languages/ruby/standard_library/11_concurrency_models/test_083_queue_sized_queue_and_close.rb

require "assertions"

A = PolyglotAssertions

ready = Queue.new
events = []
ready << proc do
  events << :first_start
  ready << proc { events << :nested }
  events << :first_end
end
ready << proc { events << :second }

ready.pop.call until ready.empty?
A.equal([:first_start, :first_end, :second, :nested], events)

# 这只是应用定义的 FIFO；Ruby 语言没有 JavaScript 式 microtask queue。
A.falsey(defined?(Microtask))

A.done
